### PR TITLE
[Configs] Add a new ConfigSanitizer component.

### PR DESCRIPTION
**Pull request by JoshLind**
Note: this PR is based on: https://github.com/aptos-labs/aptos-core/pull/7781

### Description
This PR adds a simple `ConfigSanitizer` component to the node configs. The purpose of the `ConfigSanitizer` is to validate, sanity check and sanitize all configs on node startup, to help ensure configs are well-formed and consistent (e.g., that invariants hold across node types and networks).

To do this, we: (i) load the config from disk; (ii) extract the node role from the config; (iii) extract the chain ID from the genesis blob; and (iv) validate the config according to the node role, chain ID and compilation flags. This allows us to enforce various properties about the configs. Note: we're already doing some of these verifications today, but: (i) we're missing a lot of basic checks to prevent users from shooting themselves in the foot; and (ii) whatever checks we do have are spread throughout the code, making it difficult to reason about. The `ConfigSanitizer` will at least provide us a single source of truth regarding config quality.

The PR offers two commits:
1. Add the new sanitizer trait and implementations.
2. Add unit tests to verify sanitizer correctness.

Note:
- I haven't added all the checks we may want to add (only some obvious ones), but we can extend the checks once this lands.
- My next step is to add support for fine-grained node config defaults.

### Test Plan
New and existing test infrastructure.

### CODE REVIEW

File: config/src/config/config_sanitizer.rs
Comment (davidiw) [inline config/src/config/config_sanitizer.rs:43]: let's not add redundant comments loc is not measured :P

**Review by davidiw — APPROVED** [review]

File: config/src/config/config_sanitizer.rs
Comment (JoshLind) [reply config/src/config/config_sanitizer.rs]: Lol 😆 The comments were initially there to remind me validate all sub-configs. But, I've removed them now 😄 

> loc is not measured :P
(My secret plan has been thwarted :trollface:)

File: config/src/config/config_sanitizer.rs
Comment (brianolson) [inline config/src/config/config_sanitizer.rs:114]: Is this too picky? Would someone want to turn it off (temporarily) but leave their config in place?

File: config/src/config/config_sanitizer.rs
Comment (brianolson) [inline config/src/config/config_sanitizer.rs:194]: check that `node_role.is_validator()` ?

File: config/src/config/indexer_grpc_config.rs
Comment (brianolson) [inline config/src/config/indexer_grpc_config.rs:35]: why not `impl Default for Indexer GrpcConfig {...}` ? It could set enabled=false and the other values. Then it's more clear what someone gets if they edit the .yaml and set enabled=true

File: config/src/config/logger_config.rs
Comment (brianolson) [inline config/src/config/logger_config.rs:80]: another maybe-too picky spot? It's not bad to have it disabled but configured in case of turning it back on temporarily for diagnostics?

File: types/src/chain_id.rs
Comment (JoshLind) [reply types/src/chain_id.rs:81]: Personally, I'm not a super big fan of hiding errors. I'd rather the caller know (and explicitly have to deal with it), than just return `false` and allow the code to continue. But, maybe I'm a little extreme 😄

File: config/src/config/indexer_grpc_config.rs
Comment (JoshLind) [reply config/src/config/indexer_grpc_config.rs:35]: Yeah, so this is something you can argue with @bowenyang007 and @larry-aptos about 😄 I just moved things around. I agree -- I much prefer defaulting, too.

File: config/src/config/config_sanitizer.rs
Comment (JoshLind) [reply config/src/config/config_sanitizer.rs:114]: Good question. I think `failpoints` are only enabled in specific types of test infra (e.g., forge), so I can't really imagine anyone wanting to enable the feature locally and not set the failpoints. I'm inclined to be strict, and then loosen it later (if we find a reasonable use-case)?

File: config/src/config/config_sanitizer.rs
Comment (JoshLind) [reply config/src/config/config_sanitizer.rs:194]: Aah, yeah, makes sense!

File: config/src/config/logger_config.rs
Comment (JoshLind) [reply config/src/config/logger_config.rs:80]: Potentially. @ibalajiarun actually mentioned something similar too (e.g., see here for my reasoning: https://github.com/aptos-labs/aptos-core/pull/7781). I think the best approach is to probably change the default based on the `tokio-console` feature (i.e., it's it's enabled we set the config to `Some(...)`, otherwise, we set it to `None`). And this kind of fine-grained tuning is coming soon (my next PR?), so hopefully we can get the best of both worlds once this lands.

**Review by bchocho — APPROVED** [review]
LGTM!

**Comment by JoshLind** [comment]
Landing this to unblock the quorum store checks.

**Pull request merged**
